File: lint-arwaky/archive/infrastructure/python_metrics_adapter.py
"""python_metrics_adapter — Thin adapters for Python metrics (Radon, file sizes, trends)."""

from __future__ import annotations

from ..taxonomy import (
    FilePath,
    Count,
    ResponseData,
    ResponseDataList,
    MetricsError,
    ErrorMessage,
)

import logging
import os
import json


from ..contract import IMetricsProviderPort, IPathNormalizationPort

logger = logging.getLogger(__name__)


class MetricsProvider(IMetricsProviderPort):
    """Implementation of IMetricsProvider for file and quality history."""

    def __init__(
        self, path_norm: IPathNormalizationPort, history_path=".auto_lint_history"
    ):
        self.path_norm = path_norm
        self._history_path = history_path

    async def get_line_count(self, path: FilePath) -> Count | MetricsError:
        """Returns the raw line count of a file."""
        p = str(path)
        if not os.path.isfile(p):
            return MetricsError(message=ErrorMessage(value=f"File not found: {p}"))
        try:
            with open(p, "r", encoding="utf-8") as f:
                return Count(value=len(f.readlines()))
        except Exception as e:
            return MetricsError(message=ErrorMessage(value=f"Failed to read file lines: {e}"))

    async def get_history(self) -> ResponseDataList | MetricsError:
        """Reads the raw history log."""
        if not os.path.exists(self._history_path):
            return ResponseDataList(values=[])
        history = []
        try:
            with open(self._history_path, "r", encoding="utf-8") as f:
                for line in f:
                    if line.strip():
                        history.append(ResponseData(value=json.loads(line)))
        except Exception as e:
            logger.debug("Failed to read history file %s", self._history_path)
            return MetricsError(message=ErrorMessage(value=str(e)))
        return ResponseDataList(values=history)

File: lint-arwaky/src-rust/infrastructure/python_metrics_adapter.rs
/// python_metrics_adapter — Thin adapters for Python metrics (Radon, file sizes, trends).
use crate::contract::IMetricsProviderPort;
use crate::taxonomy::{Count, ErrorMessage, FilePath, MetricsError, ResponseData, ResponseDataList};
use crate::contract::IPathNormalizationPort;
use std::sync::Arc;

pub struct MetricsProvider {
    path_norm: Arc<dyn IPathNormalizationPort>,
    history_path: String,
}

impl MetricsProvider {
    pub fn new(path_norm: Arc<dyn IPathNormalizationPort>, history_path: &str) -> Self {
        Self { path_norm, history_path: history_path.to_string() }
    }
}

#[async_trait::async_trait]
impl IMetricsProviderPort for MetricsProvider {
    async fn get_line_count(&self, path: &FilePath) -> Result<Count, MetricsError> {
        let p = &path.value;
        if !std::path::Path::new(p).is_file() {
            return Err(MetricsError::new(format!("File not found: {}", p)));
        }
        match std::fs::read_to_string(p) {
            Ok(content) => Ok(Count::new(content.lines().count() as i64)),
            Err(e) => Err(MetricsError::new(format!("Failed to read: {}", e)))
        }
    }

    async fn get_history(&self) -> Result<Vec<ResponseData>, MetricsError> {
        if !std::path::Path::new(&self.history_path).exists() {
            return Ok(Vec::new());
        }
        let mut history = Vec::new();
        if let Ok(content) = std::fs::read_to_string(&self.history_path) {
            for line in content.lines() {
                if !line.trim().is_empty() {
                    if let Ok(val) = serde_json::from_str::<serde_json::Value>(line) {
                        history.push(ResponseData::new(val));
                    }
                }
            }
        }
        Ok(history)
    }
}
